contracts/staking-vault/StakingVault.sol

```diff
@@ -0,0 +1,248 @@
+            tokenSupported[tokens[i]] = support[i];
+            unchecked { ++i; }
+        }
+
+    }
+
+    function setRewardToken(address r) external onlyOwner {
+        rewardToken = r;
+        emit RewardToken(r);
+    }
+
+    function deposit(address depositToken, address tokenToBack, uint256 amount) external {
+        require(tokenSupported[depositToken], "unsupported deposit token");
+        uint256 tokenID = convertToID(depositToken, tokenToBack);
+        if (IDToTokens[tokenID].depositToken == address(0)) {
+            sTokenPrice[tokenID] = 1e18;
+        }
+        IDToTokens[tokenID] = tokenComposition({depositToken:depositToken, tokenToBack:tokenToBack});
+        uint256 mintAmount = _amountToMint(tokenID, amount);
+        _claimReward(tokenID);
+        _mint(msg.sender, tokenID, mintAmount, "");
+        balanceStakedPerID[tokenID] += amount;
+        supplyPerID[tokenID] += mintAmount;
+        IERC20(depositToken).safeTransferFrom(msg.sender, address(this), amount);
+
+        emit Deposit(msg.sender, tokenToBack, depositToken, amount);
+    }
+
+    function withdraw(address depositToken, address tokenToBack, uint256 amount) external {
+        uint256 tokenID = convertToID(depositToken, tokenToBack);
+        _claimReward(tokenID);
+        _burn(msg.sender, tokenID, amount);
+        supplyPerID[tokenID] -= amount;
+        uint256 sendAmount = _amountToSend(tokenID, amount);
+        balanceStakedPerID[tokenID] -= sendAmount;
+        IERC20(depositToken).safeTransfer(msg.sender, sendAmount);
+
+        emit Withdraw(msg.sender, tokenToBack, depositToken, amount);
+    }
+
+    function drawOnPool(address tokenBacked, address tokenToCover, uint256 amountToCover) external onlyProtocol returns (uint256[] memory) {
+        address[] memory tokensStaked = stakingTokens;
+        uint256[] memory valuesPerToken = new uint256[](tokensStaked.length);
+        uint256 totalValue;
+        for (uint i=0;i<tokensStaked.length;) {
+            valuesPerToken[i] = priceFeed.queryReturn(tokensStaked[i], valuationToken, balanceStakedPerID[convertToID(tokensStaked[i], tokenBacked)]);
+            totalValue += valuesPerToken[i];
+            unchecked { ++i; }
+        }
+        uint256[] memory amountDrawnPerToken = new uint256[](tokensStaked.length);
+        uint256 valueOfCoverage = priceFeed.queryReturn(tokenToCover, valuationToken, amountToCover);
+        for (uint i = 0;i<tokensStaked.length;) {
+            amountDrawnPerToken[i] = priceFeed.queryReturn(valuationToken, tokensStaked[i], valueOfCoverage*valuesPerToken[i]/totalValue);
+            IERC20(tokensStaked[i]).safeTransfer(msg.sender, amountDrawnPerToken[i]);
+            unchecked {++i;}
+        }
+
+        _updatePrice(tokenBacked, amountDrawnPerToken);
+
+        emit PoolDraw(tokenBacked, tokenToCover, amountToCover);
+        return amountDrawnPerToken;
+    }
+
+    function getStoredTokenPrice(uint256 ID) external view returns (uint256) {
```

Comment:
what is the reason to have separate function when `sTokenPrice` is already public

Reply:
better written function name. It doesnt need to be added but just makes more sense than sTokenPrice.

Reply:
then make sTokenPrice private

Reply:
fixed